sound/soc/sof/topology.c

```diff
@@ -1383,67 +1390,201 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
 	kfree(tone);
 	return ret;
 }
-/*
- * Effect Topology. Only IIR equalizer is supported at this moment.
- * TODO: Need to add also FIR support and have a way to add other
- * effects and enhancements.
- */
 
-static int sof_widget_load_effect(struct snd_soc_component *scomp, int index,
-				  struct snd_sof_widget *swidget,
-				  struct snd_soc_tplg_dapm_widget *tw,
-				  struct sof_ipc_comp_reply *r)
+static int sof_effect_fir_load(struct snd_soc_component *scomp, int index,
+			       struct snd_sof_widget *swidget,
+			       struct snd_soc_tplg_dapm_widget *tw,
+			       struct sof_ipc_comp_reply *r)
+
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct snd_soc_tplg_private *private = &tw->priv;
-	struct sof_ipc_comp_eq_iir *eq;
+	struct snd_sof_control *scontrol = NULL;
+	struct sof_abi_hdr *pdata = NULL;
+	struct sof_ipc_comp_eq_fir *fir;
+	size_t ipc_size = 0, fir_data_size = 0;
 	int ret;
 
-	eq = kzalloc(sizeof(*eq), GFP_KERNEL);
-	if (!eq)
+	/* search possible initial parameters from eq controls */
+	list_for_each_entry(scontrol, &sdev->kcontrol_list, list) {
```

Comment:
@juimonen what happens if there is more than 1 control. Is that allowed?

Reply:
I mean I check that the comp_id of the control matches this components id? Can you set multiple controls from topology, isn't that quite bad already? If there are multiple controls for 1 component then this code will try to take the params from the first control...

Reply:
@juimonen yes you can have as many controls in topology as you want. Thats why we should check for sanity.